### PR TITLE
Update ServiceBus to enable live testing in sovereign clouds for multiple services

### PR DESCRIPTION
**Pull request by v-hongli1**
These changes enable search to run live tests against Public, UsGov and China.

1. Update file : IntegrationTestBase.java
    - Update the variable type of `sessionId` to `AtomicReference<String>`  for exclude interference from data of other test cases.
    - Update the way of get `sessionId` to `sessionId.get()`  for exclude interference from data of other test cases.
    - Update the way of get system parameters to `GLOBAL_CONFIGURATION.get(....)` and `TestUtils.getPropertyValue(....)`.

2. Update file : LockRenewalOperationTest.java
    - Fixed timeout exception, so remove timeout setting for test case `LockRenewalOperationTest.cancellation`

3. Update file : ServiceBusClientBuilderTest.java, ServiceBusAdministrationClientImplIntegrationTests.java, 
                        ServiceBusSharedKeyCredentialTest.java
    - Fixed endpoint url of service bus for run live test in UsGov and China cloud.

4. Update file : ServiceBusMixClientIntegrationTest.java
    - Update the way of set and get `sessionId` to `sessionId.set(UUID.randomUUID().toString());` and `sessionId.get()` for exclude interference from data of other test cases.
    - Add message receive filters for test case `ServiceBusMixClientIntegrationTest.crossEntityQueueTransactionWithReceiverSenderTest` to exclude interference from data of other test cases.

5. Update file : ServiceBusProcessorClientIntegrationTest.java, ServiceBusSessionManagerIntegrationTest.java
    - Update the way of set and get `sessionId` to `sessionId.set(UUID.randomUUID().toString());` and `sessionId.get()` for exclude interference from data of other test cases.

6. Update file : ServiceBusReceiverAsyncClientIntegrationTest.java
    - Update the way of set and get `sessionId` to `sessionId.set(UUID.randomUUID().toString());` and `sessionId.get()` for exclude interference from data of other test cases.
    - Fixed timeout exception, so remove timeout setting from test cases `receiveTwoMessagesAutoComplete`, `receiveMessageAutoComplete`, `peekMessage`, `sendScheduledMessageAndReceive`, `peekFromSequenceNumberMessage`, `peekMessagesFromSequence`, `deadLetterMessage`, `autoRenewLockOnReceiveMessage`, `receiveDeferredMessageBySequenceNumber`, `receiveFromDeadLetter`, `renewMessageLock` and private method `completeMessages` .
    - Add message receive filters for test case `receiveTwoMessagesAutoComplete` and `peekFromSequenceNumberMessage`, `peekMessages`, `deadLetterMessage` to exclude interference from date other test cases.
    - Fixed message are not always guaranteed, so update message receive mode of test case `peekFromSequenceNumberMessage` to repeat receive mode and add message receive filters to exclude interference from data other test cases.
    - Fixed exception where `sessionId` is lost unexpectedly, so add reset `sessionId` process for test case `receiveFromDeadLetter`.
    - Fixed maxMessages are not always guaranteed, so update message receive mode of test case `peekMessages` to repeat receive mode and add message receive filters to exclude interference from data other test cases and add sort to correct ordering of message list because the parallel streams used in `ServiceBusSenderAsyncClient.sendMessages(Iterable<ServiceBusMessage> messages)`. 
    
7. Update file : ServiceBusReceiverClientIntegrationTest.java
    - Update the way of set and get `sessionId` to `sessionId.set(UUID.randomUUID().toString());` and `sessionId.get()` for exclude interference from data of other test cases.
    - Fixed maxMessages are not always guaranteed, sometime, we get less than asked for, just trying two times is not enough, so update `for` loop to `while` loop and use `Thread` to control the time consumption of the verification process
    - Fixed `The operation was cancelled` exception, so add timeout setting for test case `parallelReceiveByOneSubscriber`
    
8. Update file : ServiceBusSenderAsyncClientIntegrationTest.java
    - Update the way of set and get `sessionId` to `sessionId.set(UUID.randomUUID().toString());` and `sessionId.get()` for exclude interference from data of other test cases.
   Fixed timeout exception, so add timeout setting from test case `nonSessionEntitySendMessageList`.

9. Update file : TestUtils.java
    - Fixed the way of get system parameters and get property value method.
    - Add get method for missing system parameters `AZURE_CLIENT_ID`,`AZURE_CLIENT_SECRET`,`AZURE_TENANT_ID` and global configuration.
    - Add calculation of `ShareAccessSignature`.

10. Update file : ServiceBusAdministrationAsyncClientIntegrationTest.java
    - Fixed the way of get `AzureClientId` and `AzureClientSecret` and `AzureTenantId` to `TestUtils.getAzureClientId()` and `TestUtils.getAzureClientSecret(`) and `TestUtils.getAzureTenantId()` .
    - Fixed subscription name length error, so change the length of `subscriptionName` to 50 for test case `createSubscriptionWithForwarding` . 
    - Fixed the exception that occurs when call `ServiceBusAdministrationAsyncClient.createSubscription`, so add the call of `ServiceBusAdministrationAsyncClient.createTopic` for test case `createSubscriptionWithForwarding`
    - Fixed timeout exception, so add timeout setting for test case 'createSubscriptionWithForwarding'
    - Fixed `The specified parameter (bar) in Parameters property was not used in the expression` exception, so change `sqlExpression` of  `SqlRuleFilter` to `sys.To='foo' OR sys.MessageId IS NULL` for test case `createRuleResponse`

11. Update file : test-resources.json
    - Match with the ArmTemplateParameter setting, so add the parameter `serviceBusEndpointSuffix` of the endpoint suffix.

Pipeline results:
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1501271&view=results

@benbp, @joshfree, @ramya-rao-a, @lilyjma and @ZejiaJiang  for notification.

### CODE REVIEW

**Comment by benbp** [comment]
@anuchandy can you review or @ mention an appropriate reviewer? Thanks!

**Comment by v-xuto** [comment]
@joshfree, @ramya-rao-a, @lilyjma, @anuchandy and @ZejiaJiang Could you help review this PR? Thanks a lot.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/LockRenewalOperationTest.java [hunk 1]
```diff
@@ -194,7 +194,7 @@ void cancellation() {
                 operation.close();
             })
             .expectComplete()
-            .verify(renewalPeriod);
+            .verify();
```
Comment (weidongxu-microsoft) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/LockRenewalOperationTest.java:197]: Will the remove of timeout cause test case never stop?

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/IntegrationTestBase.java [hunk 2]
```diff
@@ -61,12 +63,13 @@
...
 
     private static final String PROXY_AUTHENTICATION_TYPE = "PROXY_AUTHENTICATION_TYPE";
+    private static final Configuration GLOBAL_CONFIGURATION = TestUtils.getGlobalConfiguration();
 
     private String testName;
     private final Scheduler scheduler = Schedulers.parallel();
 
     protected static final byte[] CONTENTS_BYTES = "Some-contents".getBytes(StandardCharsets.UTF_8);
-    protected String sessionId;
+    protected AtomicReference<String> sessionId = new AtomicReference<String>();
```
Comment (weidongxu-microsoft) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/IntegrationTestBase.java:72]: I do not see why this change would "exclude interference from data of other test cases". If another thread assess it, the sessionId still changes (update of a String reference is anyway atomic).

**Review by weidongxu-microsoft — COMMENTED** [review]
A quick feedback without full context and quick read-through.

File: sdk/servicebus/tests.yml [hunk 3]
```diff
@@ -8,6 +8,6 @@ stages:
         - name: azure-messaging-servicebus
           groupId: com.azure
           safeName: azuremessagingservicebus
-      Clouds: 'Public,Canary'
+      SupportedClouds: 'Public,UsGov,China'
       EnvVars:
         AZURE_LOG_LEVEL: 2
```
[the remark this answers is not in the document]
Comment (hongli750210) [reply sdk/servicebus/tests.yml:13]: Fixed some env var redirects for AZURE_* in test.yml for the new version .

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/LockRenewalOperationTest.java (hunk 1 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/LockRenewalOperationTest.java]: With reference to your suggestion, it has been fixed in the new version.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/IntegrationTestBase.java (hunk 2 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/IntegrationTestBase.java]: With reference to your suggestion, it has been fixed in the new version.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java [hunk 4]
```diff
@@ -448,20 +454,32 @@ void cancelScheduledMessage(MessagingEntityType entityType, boolean isSessionEna
...
 
-        final ServiceBusReceivedMessage peekMessage = receiver.peekMessage().block(TIMEOUT);
-        assertNotNull(peekMessage);
+        // Message are not always guaranteed, so try many times
+        final ServiceBusReceivedMessage peekMessage = receiver.peekMessage()
+            .filter(receivedMessage -> messageId.equals(receivedMessage.getMessageId()))
+            .flatMap(receivedMessage -> {
+                countDownLatch.countDown();
+                return Mono.just(receivedMessage);
+            })
```
Comment (weidongxu-microsoft) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java:476]: Is this same as below?
```suggestion
            .map(receivedMessage -> {
                countDownLatch.countDown();
                return receivedMessage;
            })
```

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java [hunk 5]
```diff
@@ -448,20 +454,32 @@ void cancelScheduledMessage(MessagingEntityType entityType, boolean isSessionEna
...
+            .repeat(() -> countDownLatch.getCount() > 0)
+            .next()
+            .block();
+        if (!countDownLatch.await(20, TimeUnit.SECONDS)) {
+            Assertions.fail("Failed peek from sequence number message");
+        }
```
Comment (weidongxu-microsoft) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java:482]: If the line above blocks due to no message, will the test ever come to this line?

**Review by weidongxu-microsoft — COMMENTED** [review]
Added a few comments on part of the code (I still not read all).
I am not sure whether the timeout/countDownLatch works correctly in your code.

On timeout, my opinion is that
1. If you are sure it ends in one or two minutes, it does not hurt to have a timeout as e.g. 5min
2. If you are not sure how long it takes, it still need a timeout to make sure the test will not block the whole CI for hours. Just fail the test if something wrong. (for live tests, it is OK e.g. it only pass 90% as long as we know that 10% is not due to SDK problem)

You can wait for feedback from @anuchandy or @conniey before further changes.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java [hunk 6]
```diff
@@ -325,18 +328,14 @@ public void testConnectionWithAzureNameKeyCredential() {
             .credential("",
                 new AzureNamedKeyCredential(sharedAccessKeyName, sharedAccessKey)));
 
-        assertThrows(IllegalArgumentException.class, () -> new ServiceBusClientBuilder()
-            .credential(fullyQualifiedNamespace,
-                new AzureNamedKeyCredential(sharedAccessKeyName, sharedAccessKey)));
```
[the remark this answers is not in the document]
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java:330]: Expected java.lang.IllegalArgumentException to be thrown, but nothing was thrown, so removal this 2 assertions.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java (hunk 5 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java:482]: With reference to your suggestion, fixed in the new version.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java [hunk 7]
```diff
@@ -511,40 +531,112 @@ void peekMessages(MessagingEntityType entityType, boolean isSessionEnabled) {
...
+                receivedMessages.forEach(actualMessage -> checkCorrectMessage.accept(actualMessage, 7));
+            }));
+
+            threadList.parallelStream().forEach(t -> {
+                t.start();
+                try {
+                    t.join(OPERATION_TIMEOUT.toMillis());
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
```
Comment (weidongxu-microsoft) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java:624]: Which thread does this join happen?

I assume `join` should happen at the thread of the test case. But I am not sure which thread runs this in `parallelStream`.

And what happens if `join` pass the timeout?

**Review by weidongxu-microsoft — COMMENTED** [review]
I am not very comfortable (it's complicated and I found some bug but I won't know if I found all) about the code on thread and join and timeout.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java (hunk 7 above)
Comment (weidongxu-microsoft) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java:624]: Or should you use `ExecutorService` instead list of threads?

Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java:624]: 1. When we invoke the join() method on test thread, the test thread goes into a waiting state. It remains in a waiting state until the referenced thread terminates. If the referenced thread is blocked or is taking too long to process. `The operation was canceled` exception will occur on the test thread. 
2.  join() with Timeout means to waits at most millis milliseconds for this thread to die.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java]: The two removed assertions have complete parameters, so `Expected java.lang.IllegalArgumentException to be thrown, but nothing was thrown` exception will occur when executing these two assertions, so such assertions must be split and refined. the split and refined assertions have been fixed in the new version.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java [hunk 8]
```diff
@@ -84,10 +96,48 @@
      * @return The namespace connection string.
      */
     public static String getConnectionString(boolean withSas) {
+        String connectionString = getPropertyValue("AZURE_SERVICEBUS_NAMESPACE_CONNECTION_STRING");
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java:99]: Is there a reason we're copying this entire logic instead of using the ServiceBusSharedKeyCredential and generating the signature from that?

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/administration/ServiceBusAdministrationAsyncClientIntegrationTest.java [hunk 9]
```diff
@@ -406,6 +410,15 @@ void createSubscriptionWithForwarding(HttpClient httpClient) {
             .setForwardDeadLetteredMessagesTo(forwardToTopic);
 
         // Act & Assert
+        if (!interceptorManager.isPlaybackMode()
+            && !client.getTopicExists(topicName).block()) {
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/administration/ServiceBusAdministrationAsyncClientIntegrationTest.java:414]: instead of making two network calls you can just handle the ResourceExistsException.

```java
        if (!interceptorManager.isPlaybackMode()) {
            client.createTopic(topicName)
                .onErrorResume(ResourceExistsException.class, error -> Mono.empty())
                .block(TIMEOUT);

            client.createTopic(forwardToTopic)
                .onErrorResume(ResourceExistsException.class, error -> Mono.empty())
                .block(TIMEOUT);
        }
```

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java [hunk 10]
```diff
@@ -377,4 +395,8 @@ private static URI getUri(String endpointFormat, String namespace, String domain
                 "Invalid namespace name: %s", namespace), exception);
         }
     }
+
+    public static String getEndPoint() {
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java:399]: Why a helper method that just returns a property? Replace with TestUtils.getEndpoint();

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java [hunk 11]
```diff
@@ -96,7 +146,11 @@ public static String getConnectionString(boolean withSas) {
      * @return The fully qualified domain name for the service bus resource.
      */
     public static String getFullyQualifiedDomainName() {
-        return System.getenv("AZURE_SERVICEBUS_FULLY_QUALIFIED_DOMAIN_NAME");
+        return getPropertyValue("AZURE_SERVICEBUS_FULLY_QUALIFIED_DOMAIN_NAME");
+    }
+
+    public static String getEndPoint() {
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java:152]: rename to: getEndpoint()

File: sdk/servicebus/test-resources.json [hunk 12]
```diff
@@ -68,7 +71,7 @@
   "resources": [
     {
       "type": "Microsoft.ServiceBus/namespaces",
-      "apiVersion": "2021-05-01",
+      "apiVersion": "2021-01-01-preview",
```
Comment (conniey) [inline sdk/servicebus/test-resources.json:74]: Is there a reason we're using a different API version that is older?

Comment (hongli750210) [reply sdk/servicebus/test-resources.json:74]: When using 2021-05-01 api version, failed to deploy test resource in weekly pipeline. More details please check [issue#26058](https://github.com/Azure/azure-sdk-for-java/issues/26058).

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java (hunk 11 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java:152]: It has been verified that this getter method can be replaced by `TestUtils.getEndPoint()`, now fixed in the new version.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java (hunk 10 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java]: It has been verified that this getter method can be replaced by `TestUtils.getEndPoint()`, now fixed in the new version.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/administration/ServiceBusAdministrationAsyncClientIntegrationTest.java (hunk 9 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/administration/ServiceBusAdministrationAsyncClientIntegrationTest.java]: According to your suggestion, now fixed in the new version.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java (hunk 8 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java:99]: The connection string of `ServiceBus` has `Endpoint=...;SharedAccessKeyName=...;SharedAccessKey=...` and `Endpoint=...;SharedAccessKeyName=...;SharedAccessKey=...;EntityPath=...` these two formats, in order to the extensibility of test cases and the generality of the generated sas connection string, so copying this entire logic.

**Comment by v-xuto** [comment]
@conniey We have updated by your comments. Please re-review this PR.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java [hunk 13]
```diff
@@ -37,7 +37,7 @@
 
 class ServiceBusClientBuilderTest extends IntegrationTestBase {
     private static final String NAMESPACE_NAME = "dummyNamespaceName";
-    private static final String DEFAULT_DOMAIN_NAME = "servicebus.windows.net/";
+    private static final String DEFAULT_DOMAIN_NAME = TestUtils.getEndPoint().substring(1) + "/";
```
Comment (ki1729) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java:40]: It would be better to definte the domain name as a seperate entity instead of relying on the endpoint value.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java (hunk 11 above)
Comment (ki1729) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java:152]: Looks like this change regressed

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java (hunk 13 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java:40]: @ki1729  The endpoint suffix of Service bus is different on UsGov and Azure China cloud.  So it is more appropriate to relying on the endpoint value  . For details, please refer to Issue#[23646](https://github.com/Azure/azure-sdk-for-java/issues/23646)

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java (hunk 11 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/TestUtils.java]: Sorry for missing this fix in the previous revision, now fixed in the new version.

**Review by ki1729 — APPROVED** [review]
Looks good Hong!

**Comment by joshfree** [comment]
@ki1729 could you help shepherd this PR in as you have time?

**Comment by ki1729** [comment]
On it @joshfree

**Comment by hongli750210** [comment]
@ki1729 
The Service Bus pipeline weekly test has been successed after renew, For more details please check [here](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1467029&view=results)

**Comment by benbp** [comment]
Thanks @hongli750210!

**Comment by ki1729** [comment]
@hongli750210 are we waiting for any other changes? If not can you please merge this PR?

**Comment by v-xuto** [comment]
@benbp Now, the result of the service bus weekly pipeline is `Pass`, for more details please check here:
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1501271&view=results

Besides, Ki1729 have approved this PR. Do you have any questions about this PR? If not, please help merge this PR.
